File: Scala/data_structures/src/main/scala-2.12/introduction/CheckBrackets.scala
package introduction

object CheckBrackets {
  import scala.annotation.tailrec

  type Stack[T] = List[T]

  private val OpeningBrackets: Set[Char] = Set('(', '[', '{')
  private val ClosingBrackets: Set[Char] = Set(')', ']', '}')

  final case class OpenedBracket(bracket: Char, position: Int)

  private def doBracketsMatch(openingBracket: Char, closingBracket: Char): Boolean =
    (openingBracket == '(' && closingBracket == ')') ||
      (openingBracket == '[' && closingBracket == ']') ||
      (openingBracket == '{' && closingBracket == '}')

  private def retrieveFailedOpeningIndexFromStack(stack: Stack[OpenedBracket]): Option[Int] = stack match {
    case Nil => None
    case OpenedBracket(_, ix) :: _ => Some(ix)
  }

  def findIndexOfNonMatchingBracket(string: List[Char]): Option[Int] = {
    @tailrec
    def loop(stack: Stack[OpenedBracket],
             currentString: List[(Char, Int)]): (Stack[OpenedBracket], Option[Int]) = currentString match {
      case Nil => (stack, None)
      case (letter, ix) :: rest =>
        if (OpeningBrackets.contains(letter)) loop(OpenedBracket(letter, ix) :: stack, rest)
        else if (ClosingBrackets.contains(letter)) {
          stack match {
            case Nil => (stack, Some(ix))
            case OpenedBracket(openingBracket, _) :: stackTail =>
              if (doBracketsMatch(openingBracket, letter)) loop(stackTail, rest)
              else (stack, Some(ix))
          }
        } else loop(stack, rest)
    }

    val (remainingStack, failedIndex): (Stack[OpenedBracket], Option[Int]) = loop(Nil, string.zipWithIndex)
    failedIndex match {
      case None => retrieveFailedOpeningIndexFromStack(remainingStack)
      case _ => failedIndex
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val string: List[Char] = reader.next().toList
    findIndexOfNonMatchingBracket(string) match {
      case None => println("Success")
      case Some(ix) => println(ix + 1)
    }
  }
}

File: OCaml/data_structures/introduction/check_brackets.ml
let opening_brackets = ['('; '['; '{']


let closing_brackets = [')'; ']' ; '}']


type opened_bracket = { bracket : char; position : int }


let do_brackets_match opening_bracket closing_bracket =
    (opening_bracket = '(' && closing_bracket = ')') ||
    (opening_bracket = '[' && closing_bracket = ']') ||
    (opening_bracket = '{' && closing_bracket = '}')


let retrieve_failed_opening_index = function
    | [] -> None
    | { position } :: _ -> Some position


let find_index_of_non_matching_bracket line =
    let length = String.length line in
    let rec loop stack ix =
        if ix = length then stack, None
        else
            let character = String.get line ix in
            if List.mem character opening_brackets then loop ({ bracket = character; position = ix } :: stack) (ix + 1)
            else if List.mem character closing_brackets then
                match stack with
                    | [] -> stack, Some ix
                    | { bracket = opening_bracket } :: stack_tail ->
                        if do_brackets_match opening_bracket character then loop stack_tail (ix + 1)
                        else stack, Some ix
            else loop stack (ix + 1) in
    let remaining_stack, failed_index = loop [] 0 in
    match failed_index with
        | None -> retrieve_failed_opening_index remaining_stack
        | _ -> failed_index


let () =
    let line = read_line () in
    match find_index_of_non_matching_bracket line with
        | None -> print_endline "Success"
        | Some ix -> print_int (ix + 1); print_newline()
